File: src/main/java/com/learnershi/rclasssocket/repository/ClassRoomUserRepositoryImpl.java
package com.learnershi.rclasssocket.repository;

import com.learnershi.rclasssocket.entity.User;
import org.springframework.stereotype.Component;

import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Component
public class ClassRoomUserRepositoryImpl implements ClassRoomUserRepository {
    private static Map<String, Map<String, User>> classroomUserMap;
    private ClassRoomRepository classRoomRepository;

    public ClassRoomUserRepositoryImpl() {
        if (classroomUserMap == null) {
            classroomUserMap = new ConcurrentHashMap<>();
        }
    }

    public void addUser(String classRoomId, User user) {
        Map<String, User> userMap = classroomUserMap.get(classRoomId);
        if (userMap == null) {
            userMap = new ConcurrentHashMap<>();
        }
        userMap.put(user.getSeq(), user);
        classroomUserMap.put(classRoomId, userMap);
    }

    public User removeUser(String classRoomId, String userSeq) {
        Map<String, User> userMap = classroomUserMap.get(classRoomId);
        User user = null;
        if (userMap != null) {
            user = userMap.get(userSeq);
            userMap.remove(userSeq);
        }
        return user;
    }

    public Collection<User> findByClassRoomId(String classRoomId) {
        if (classroomUserMap.get(classRoomId) == null) {
            return Collections.EMPTY_SET;
        }
        return classroomUserMap.get(classRoomId).values();
    }


    public User findUserByClassRoomIdAndUserSeq(String classRoomId, String userSeq) {
        Map<String, User> userMap = classroomUserMap.get(classRoomId);
        if (userMap == null) {
            return null;
        }
        return userMap.get(userSeq);
    }
}

File: src/main/java/com/learnershi/rclasssocket/repository/ClassRoomUserRepositoryImpl.kt
package com.learnershi.rclasssocket.repository

import com.learnershi.rclasssocket.entity.User
import org.springframework.stereotype.Component
import java.util.concurrent.ConcurrentHashMap

@Component
class ClassRoomUserRepositoryImpl : ClassRoomUserRepository {
    init {
        if (classroomUserMap == null) {
            classroomUserMap = ConcurrentHashMap()
        }
    }

    override fun addUser(classRoomId: String, user: User) {
        var userMap = classroomUserMap!![classRoomId]
        if (userMap == null) {
            userMap = ConcurrentHashMap()
        }
        userMap[user.seq] = user
        classroomUserMap!![classRoomId] = userMap
    }

    override fun removeUser(classRoomId: String, userSeq: String): User? {
        val userMap = classroomUserMap!![classRoomId]
        var user: User? = null
        if (userMap != null) {
            user = userMap[userSeq]
            userMap.remove(userSeq)
        }
        return user
    }

    override fun findByClassRoomId(classRoomId: String): Collection<User?> {
        return if (classroomUserMap!![classRoomId] == null) {
            emptySet()
        } else classroomUserMap!![classRoomId]!!.values
    }

    override fun findUserByClassRoomIdAndUserSeq(classRoomId: String, userSeq: String): User? {
        val userMap = classroomUserMap!![classRoomId]
            ?: return null
        return userMap[userSeq]
    }

    companion object {
        private var classroomUserMap: MutableMap<String?, MutableMap<String, User?>?>? = null
    }

}